src/librustc_resolve/lib.rs

```diff
@@ -3140,19 +3165,23 @@ impl<'a> Resolver<'a> {
             if qself.position == 0 {
                 // FIXME: Create some fake resolution that can't possibly be a type.
                 return Some(PathResolution::with_unresolved_segments(
-                    Def::Mod(DefId::local(CRATE_DEF_INDEX)), path.len()
+                    Def::Mod(DefId::local(CRATE_DEF_INDEX)), path.segments.len()
                 ));
             }
             // Make sure `A::B` in `<T as A>::B::C` is a trait item.
-            let ns = if qself.position + 1 == path.len() { ns } else { TypeNS };
-            let res = self.smart_resolve_path_fragment(id, None, &path[..qself.position + 1],
+            let ns = if qself.position + 1 == path.segments.len() { ns } else { TypeNS };
+            let path = ResolvePath {
+                segments: &path.segments[..qself.position + 1],
+                source: Some(id),
+            };
+            let res = self.smart_resolve_path_fragment(id, None, &path,
                                                        span, PathSource::TraitItem(ns));
             return Some(PathResolution::with_unresolved_segments(
-                res.base_def(), res.unresolved_segments() + path.len() - qself.position - 1
+                res.base_def(), res.unresolved_segments() + path.segments.len() - qself.position - 1
             ));
         }
 
-        let result = match self.resolve_path(&path, Some(ns), true, span, Some(id)) {
+        let result = match self.resolve_path(&path, Some(ns), true, span) {
```

Comment:
the path here is incorrect and probably is the cause of the bug

Reply:
Because the source is incorrect? Redefining the path as`let path = ResolvePath { segments: path.segments, source: Some(id) };` before that line has no effect, however.

Reply:
oh sorry, I didn't notice the scopes here